Fixup for Algebraic Relationship Managers

permcody opened the pull request:
<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->


This should be mergeable if all the tests pass but I'm certainly not finished. I'm planning to add several more tests and also make it so that you can "demote" each RM if you don't need the algebraic coupling.

permcody committed:
Fixup for Algebraic Relationship Managers

permcody commented:
Oh wait, I lied - this still requires a libMesh update...

permcody committed:
Allow Developers to restrict (demote) the functionality of RMs

permcody committed:
Rename Ghost objects to RM objects

permcody committed:
Renaming and reworking RelationshipManager Tests

permcody committed:
Adding the ability to demote an RM

Also adding new tests

closes #11981

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/236769/) on bcfeaae wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/12194/site)

This comment will be updated on new commits.

permcody committed:
Fixup tests and restrict 3D partition problem to Mac only for now

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit 8516b0b93ad357ea9ec904d9021c18cf939e3bcb
+Subproject commit feecc8fdd9ce92d967f381117844eb8f63785a7a
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

permcody committed:
Grain Tracker Enhancements

permcody committed:
Add new Distributed Polycrstal test and document remaining GrainTracker tests

closes #12216

permcody committed:
Make RelationshipManager comparable

Add the ability to remove MooseObjects.

permcody committed:
Adding requirements and design to reconstruction tests

permcody committed:
Handle degenerative bounding box condition

permcody committed:
Make the grain tracker non-recursive

closes #11985

permcody committed:
Allow grain tracker to start with zero grains

closes #12200

moosebuild commented:
Job [App tests](https://civet.inl.gov/job/240506/) on 5d14634 : invalidated by @permcody

Grizzly test appears to be broken

permcody committed:
Adding a "try-harder" loop to the GrainTracker

This loop is intended to look at the previous state to match up
new grains in an attempt to properly classify them as split grains
as opposed to a new grain (nucleated). There's more to do here, but
this is a reasonable start that will work for a number of cases.

What's missing is a good way to test this. I'll try to work something
up, but it's a very difficult case to make happen.

closes #12236

permcody committed:
Allow grain tracker to start with zero grains

closes #12200

fdkong commented on modules/phase_field/test/tests/grain_tracker_test/distributed_poly_ic.i:
```diff
@@ -0,0 +1,271 @@
+[BCs]
+  # Boundary Condition block
+[]
+
```
What boundary conditions are we using here? `Neumann`? We may not need this block.

permcody replied:
No flux - this is completely fine for GrainGrowth type simulations.

fdkong commented on modules/phase_field/test/tests/grain_tracker_test/distributed_poly_ic.i:
```diff
@@ -0,0 +1,271 @@
+[Problem]
+  type = FEProblem
+[]
+
```
Do not we use `FEProblem` by default?

fdkong commented on test/src/userobjects/ElemSideNeighborLayersTester.C:
```diff
@@ -63,13 +66,20 @@ GhostUserObject::execute()
+  else if (field_name == "ghosted")
+    return _ghost_data.find(element_id) != _ghost_data.end();
+
+  return -1;
```
Return `-1` as an `unsigned long `?

permcody replied:
Good catch - that's a bad idea!

fdkong commented on test/src/userobjects/ElemSideNeighborLayersTester.C:
```diff
@@ -63,13 +66,20 @@ GhostUserObject::execute()
+  if (field_name == "evaluable")
+    return _evaluable_data.find(element_id) != _evaluable_data.end();
+  else if (field_name == "ghosted")
+    return _ghost_data.find(element_id) != _ghost_data.end();
```
Return `boolean` as an `unsigned long`?

permcody replied:
Now this is OK - Booleans fit into other interger types in a well-defined way. We are just going to use them to display "binary" type fields.

fdkong commented:
Looks good for me. Just several minor typo.

permcody commented on modules/phase_field/examples/grain_growth/grain_growth_3D.i:
```diff
@@ -9,42 +9,48 @@
 [GlobalParams]
   # Parameters used by several kernels that are defined globally to simplify input file
-  op_num = 9 # Number of order parameters used
+  op_num = 15 # Number of order parameters used
```
Right, now that the initial condition has been fixed - it's seeing a lot more conflicts. It turns out that for the 25 grains, we now need 15 order parameters instead of 9 like I was getting away with before. The upshot is that we don't have any merging grains on the first timestep!

permcody committed:
Addressing reviewer comments
